### PR TITLE
BHBC-2166: Enhancements to EML Function to Support Project/Survey EML Generation

### PR DESCRIPTION
**Pull request by curtisupshall**
# Overview
The current EML code builds an EML json object for a project and all of its surveys. We need to modify this code so that we have 2 distinct functions, one that builds an EML object for a project and one that builds an EML object for a survey.

## Links to Jira tickets

 - https://quartech.atlassian.net/browse/BHBC-2166

## Description of relevant changes

 - `eml-service` is enhanced to support a function for building EML for a single project, which includes references to any surveys under the related projects section.

 - `eml-service` is enhanced to support a function for building EML for a single survey, which includes a reference to the parent project under the related projects section.

 - All “system” node attributes are to be left as empty strings, as this will be populated by BioHub. Example:

```xml
<!-- before -->
<dataset system="https://sims.nrs.gov.bc.ca" id="a3659666-0879-4b4f-aedb-467891db198a">

<!-- after -->
<dataset system="" id="a3659666-0879-4b4f-aedb-467891db198a">
```

 - All “TaxonID” node “provider” attributes are to be left as empty strings, and the value of the node now reflects the primary key of taxonomic code. Example:

```xml
<!-- before -->
<taxonId provider="https://sims.nrs.gov.bc.ca">M-ALAM</taxonId>

<!-- after -->
<taxonId provider="">2065</taxonId>
```

 - Additionally, removed the following nodes:
   - `eml.access` node
   - `eml.dataset.metadataProvider`
   - `additionalMetadata` nodes associated with `proprietaryDataCategory`
   - `additionalMetadata` nodes associated with `proprietorName`
   - `additionalMetadata` nodes associated with `proprietaryDataCategoryRationale`
   - `additionalMetadata` nodes associated with `dataSharingAgreementRequired`

## How to Test
1. Create a project and a survey.
2. Visit a survey.
3. Click Submit Data
4. Check the API logs. The project EML and survey EML strings should be printed

### CODE REVIEW

File: api/src/services/platform-service.ts [hunk 1]
```diff
@@ -112,9 +112,11 @@ export class PlatformService extends DBService {
         return;
       }
 
-      const emlService = new EmlService({ projectId: projectId }, this.connection);
+      const emlService = new EmlService(this.connection);
+      const emlPackage = await emlService.buildProjectEmlPackage({ projectId });
+      const emlString = emlPackage.toString();
 
-      const emlString = await emlService.buildProjectEml();
+      defaultLog.debug({ label: 'submitDwCAMetadataPackage', emlString });
```
Comment (curtisupshall) [inline api/src/services/platform-service.ts]: Will remove logging `emlString` when done testing!

File: api/src/services/eml-service.ts [hunk 2]
```diff
@@ -87,65 +303,113 @@ export type BuildProjectEMLOptions = {
...
+        .withDataset(await this._buildEmlDatasetSection(packageId, projectSource.projectData))
 
-    this._packageId = options.packageId;
+        // Build EML->Dataset->Project field
+        .withProject(this._buildProjectEmlProjectSection(projectSource.projectData))
 
-    this.projectId = options.projectId;
+        // Build EML->Dataset->Project->AdditionalMetadata field
+        .withAdditionalMetadata(await this._getProjectAdditionalMetadata(projectSource))
+        .withAdditionalMetadata(this._getSurveyAdditionalMetadata(projectSource.surveys))
```
Comment (NickPhura) [inline api/src/services/eml-service.ts]: I think we can remove the survey additional metadata from the project dataset, and vice versa for the survey dataset below.

Comment (NickPhura) [reply api/src/services/eml-service.ts]: Though, its all a bit up for interpretation and how we want to use it, so might need further discussion with Robin or Charlie

My assumption is that we only include other dataset info in the related dataset section.

Though, now that I say that, the additional metadata section does have that `describes` field, to inform the EML which items this meta is for. So maybe its not incorrect to include it... Hmm..

Comment (curtisupshall) [reply api/src/services/eml-service.ts]: Yeah this was a big question mark for me. Good point

File: api/src/services/eml-service.ts [hunk 3]
```diff
@@ -49,36 +52,249 @@ type SurveyObjectWithAttachments = SurveyObject & {
...
+    return this;
+  }
+
+  /**
+   * Compiles the EML package
+   *
+   * @return {*}  {EmlPackage}
+   * @memberof EmlPackage
+   */
+  build(): EmlPackage {
```
Comment (NickPhura) [inline api/src/services/eml-service.ts:206]: Nice builder pattern!

File: api/src/services/eml-service.ts [hunk 4]
```diff
@@ -49,36 +52,254 @@ type SurveyObjectWithAttachments = SurveyObject & {
...
-type Cache = {
-  projectData?: IGetProject;
-  surveyData?: SurveyObjectWithAttachments[];
-  projectAttachmentData?: GetProjectAttachmentsData;
-  projectReportAttachmentData?: GetProjectReportAttachmentsData;
-  codes?: IAllCodeSets;
+type ProjectMetadataSource = {
+  projectData: IGetProject;
+  projectAttachmentsData: GetProjectAttachmentsData;
+  projectReportAttachmentsData: GetProjectReportAttachmentsData;
```
Comment (NickPhura) [inline api/src/services/eml-service.ts]: I can confirm with Charlie during our meeting this afternoon at 2, but I suspect we will be removing all project/survey attachment/report related information from the EML files, now that we have that information being sent separately. I forgot we even had it in here originally tbh.

File: api/src/services/eml-service.ts (hunk 2 above)
Comment (curtisupshall) [reply api/src/services/eml-service.ts]: Resolved: as discussed, we'll maintain both project and survey additional metadata on both projects and surveys. Granted, there are _currently_ no additional metadata fields provided for surveys.

File: api/src/services/eml-service.ts (hunk 4 above)
Comment (curtisupshall) [reply api/src/services/eml-service.ts]: As per dicusssion with Charlie, we're removing attachment/report information in both project and survey EML. Acceptance criteria have been changed accordingly

**Comment by NickPhura** [comment]
Looks like just some linting issues, and a few sonar cloud items.

Otherwise looks good

**Comment by curtisupshall** [comment]
> Looks like just some linting issues, and a few sonar cloud items.
> 
> Otherwise looks good

Roger that. Just adding a bit more test coverage

**Review by NickPhura — APPROVED** [review]

**Pull request merged**